abi-versions/vNEXT/README.md

```diff
@@ -0,0 +1,1471 @@
+
+Returns `Ok` on success.
+
+Returns `BadArgument` for unknown `context_id`.
+
+#### `proxy_context_finalize`
+
+* params:
+  - none
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Indicates to the host environment that the Wasm module finished processing
+current context. It should be used after previously returning `false` in
+`proxy_on_context_finalize`.
+
+Returns `Ok` on success.
+
+
+## Configuration
+
+### Callbacks optionally exposed by the Wasm module
+
+#### `proxy_on_vm_start`
+
+* params:
+  - `i32 (uint32_t) vm_id`
+  - `i32 (size_t) vm_configuration_size`
+* returns:
+  - `i32 (bool) is_success`
+
+Called when the host environment starts WebAssembly Virtual Machine (WasmVm).
+
+Its configuration (of size `vm_configuration_size`) can be retrieved using
+`proxy_get_buffer`.
+
+Return value (`is_success`) indicates that configuration was processed correctly
+and that WasmVm started successfully.
+
+#### `proxy_on_plugin_start`
+
+* params:
+  - `i32 (uint32_t) plugin_id`
+  - `i32 (size_t) plugin_configuration_size`
+* returns:
+  - `i32 (bool) is_success`
+
+Called when the host environment starts plugin.
+
+Its configuration (of size `plugin_configuration_size`) can be retrieved using
+`proxy_get_buffer`.
+
+Return value (`is_success`) indicates that configuration was processed correctly
+and that plugin started successfully.
+
+
+## TCP/UDP/QUIC stream (L4) extensions
+
+Note: `downstream` means the connection between client and proxy, and `upstream`
+       means the connection between proxy and backend (aka “next hop”).
+
+Note: The same unique context identifier (`stream_id`) is shared by a pair of
+      `downstream` and `upstream` connections.
+
+### Callbacks optionally exposed by the Wasm module
+
+#### `proxy_on_new_connection`
+
+* params:
+  - `i32 (uint32_t) stream_id`
+* returns:
+  - `i32 (proxy_action_t) next_action`
+
+Called on a new connection (`stream_id`).
+
+Return value (`next_action`) instructs the host environment what action to take:
+- `Continue` means that this connection should be accepted.
+- `Done` means that this connection should be accepted and that this extension
+  shouldn't be called again for the current stream.
+- `Pause` means that further processing should be paused until
+  `proxy_resume_stream` is called.
+- `Close` means that this connection should be rejected.
+
+#### `proxy_on_downstream_data`
+
+* params:
+  - `i32 (uint32_t) stream_id`
+  - `i32 (size_t) data_size`
+  - `i32 (bool) end_of_stream`
+* returns:
+  - `i32 (proxy_action_t) next_action`
+
+Called for each data chunk received from downstream (`stream_id`).
+
+Data (of size `data_size`) can be retrieved using `proxy_get_buffer`.
+
+Return value (`next_action`) instructs the host environment what action to take:
+- `Continue` means that data chunk should be forwarded upstream.
+- `EndStream` means that data chunk should be forwarded upstream and that this
+  connection should be closed afterwards.
```

Comment:
What does this mean in terms of half duplex? It's a bit odd to me that this is a return code. Would it be better to have this happen via a callback on the connection?